apiserver/facades/client/cloud/cloud.go

```diff
@@ -379,9 +414,125 @@ func (api *CloudAPI) Credential(args params.Entities) (params.CloudCredentialRes
+	// TODO (anastasiamac 2018-02-16) This is needed as facade v2 has CAAS-specific calls
+	// but we need to bump facade version and make v3 available.
+	if !featureflag.Enabled(feature.CAAS) {
+		return errors.NotSupportedf("AddCloud call")
+	}
 	err := api.backend.AddCloud(common.CloudFromParams(cloudArgs.Name, cloudArgs.Cloud))
 	if err != nil {
 		return err
 	}
 	return nil
 }
+
+// CredentialContents returns the specified cloud credentials,
+// including the secrets if requested.
+// If no specific credential name/cloud was passed in, all credentials for this user
+// are returned.
+// Only credential owner can see its contents as well as what models use it.
+// Controller admin has no special superpowers here and is treated the same as all other users.
+func (api *CloudAPIV3) CredentialContents(args params.CredentialContentRequestParam) (params.CredentialContentResults, error) {
+	// Helper to look up and cache credential schemas for clouds.
+	schemaCache := make(map[string]map[cloud.AuthType]cloud.CredentialSchema)
+	credentialSchemas := func(cloudName string) (map[cloud.AuthType]cloud.CredentialSchema, error) {
+		if s, ok := schemaCache[cloudName]; ok {
+			return s, nil
+		}
+		cloud, err := api.backend.Cloud(cloudName)
+		if err != nil {
+			return nil, err
+		}
+		provider, err := environs.Provider(cloud.Type)
+		if err != nil {
+			return nil, err
+		}
+		schema := provider.CredentialSchemas()
+		schemaCache[cloudName] = schema
+		return schema, nil
+	}
+
+	// Helper to parse state.Credential into an expected result item.
+	stateIntoParam := func(credential state.Credential, includeSecrets bool) params.CredentialContentResult {
+		schemas, err := credentialSchemas(credential.Cloud)
+		if err != nil {
+			return params.CredentialContentResult{Error: common.ServerError(err)}
+		}
+		secrets := map[string]string{}
+		attrs := map[string]string{}
+		// Filter out the secrets.
+		if s, ok := schemas[cloud.AuthType(credential.AuthType)]; ok {
+			for _, attr := range s {
+				if value, exists := credential.Attributes[attr.Name]; exists {
+					if attr.Hidden {
+						secrets[attr.Name] = credential.Attributes[attr.Name]
+						continue
+					}
+					attrs[attr.Name] = value
+				}
+			}
+		}
+		info := params.ControllerCredentialInfo{
+			Content: params.CredentialContent{
+				Name:       credential.Name,
+				AuthType:   credential.AuthType,
+				Attributes: attrs,
+				Cloud:      credential.Cloud,
+			},
+		}
+		if includeSecrets {
+			info.Content.Secrets = secrets
+		}
+
+		// get models
+		tag, err := credential.CloudCredentialTag()
+		if err != nil {
+			return params.CredentialContentResult{Error: common.ServerError(err)}
+		}
+
+		models, err := api.backend.CredentialModelsAndOwnerAccess(tag)
+		if err != nil && !errors.IsNotFound(err) {
+			return params.CredentialContentResult{Error: common.ServerError(err)}
+		}
+		info.Models = make([]params.ModelAccess, len(models))
```

Comment:
This is ok because if not found len(models) is 0, but typically we'll wrap the logic here in an if err == nil block

Reply:
I know... I did this on purpose as I am already checking above and will get a NotFound when len(models) == 0 from state...